specs/_features/peerdas/polynomial-commitments-sampling.md

```diff
@@ -0,0 +1,519 @@
+| - | - | - |
+| `PolynomialCoeff` | `Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB]` | A polynomial in coefficient form |
+| `Cell` | `Vector[BLSFieldElement, FIELD_ELEMENTS_PER_CELL]` | The unit of blob data that can come with their own KZG proofs |
+| `CellID` | `uint64` | Cell identifier |
+
+## Constants
+
+| Name | Value | Notes |
+| - | - | - |
+
+## Preset
+
+### Cells
+
+Cells are the smallest unit of blob data that can come with their own KZG proofs. Samples can be constructed from one or several cells (e.g. an individual cell or line).
+
+| Name | Value | Description |
+| - | - | - |
+| `FIELD_ELEMENTS_PER_CELL` | `uint64(64)` | Number of field elements in a cell |
+| `BYTES_PER_CELL` | `FIELD_ELEMENTS_PER_CELL * BYTES_PER_FIELD_ELEMENT` | The number of bytes in a cell |
+| `CELLS_PER_BLOB` | `((2 * FIELD_ELEMENTS_PER_BLOB) // FIELD_ELEMENTS_PER_CELL)` | The number of cells in a blob |
+
+### Crypto
+
+| Name | Value | Description |
+| - | - | - |
+| `ROOT_OF_UNITY_EXTENDED` | `pow(PRIMITIVE_ROOT_OF_UNITY, (BLS_MODULUS - 1) // int(FIELD_ELEMENTS_PER_BLOB * 2), BLS_MODULUS)` | Root of unity of order FIELD_ELEMENTS_PER_BLOB * 2 over the BLS12-381 field |
+| `ROOTS_OF_UNITY_EXTENDED` | `([pow(ROOT_OF_UNITY_EXTENDED, i, BLS_MODULUS) for i in range(FIELD_ELEMENTS_PER_BLOB * 2)])` | Roots of unity of order FIELD_ELEMENTS_PER_BLOB * 2 over the BLS12-381 field |
+| `ROOT_OF_UNITY_REDUCED` | `pow(PRIMITIVE_ROOT_OF_UNITY, (BLS_MODULUS - 1) // int(CELLS_PER_BLOB), BLS_MODULUS)` | Root of unity of order CELLS_PER_BLOB over the BLS12-381 field |
+| `ROOTS_OF_UNITY_REDUCED` | `([pow(ROOT_OF_UNITY_REDUCED, i, BLS_MODULUS) for i in range(CELLS_PER_BLOB)])` | Roots of unity of order CELLS_PER_BLOB over the BLS12-381 field |
+
+## Helper functions
+
+### Linear combinations
+
+#### `g2_lincomb`
+
+```python
+def g2_lincomb(points: Sequence[KZGCommitment], scalars: Sequence[BLSFieldElement]) -> Bytes96:
+    """
+    BLS multiscalar multiplication in G2. This function can be optimized using Pippenger's algorithm and variants.
+    """
+    assert len(points) == len(scalars)
+    result = bls.Z2()
+    for x, a in zip(points, scalars):
+        result = bls.add(result, bls.multiply(bls.bytes96_to_G2(x), a))
+    return Bytes96(bls.G2_to_bytes96(result))
+```
+
+### FFTs
+
+#### `_fft_field`
+
+```python
+def _fft_field(vals: Sequence[BLSFieldElement],
+               roots_of_unity: Sequence[BLSFieldElement]) -> Sequence[BLSFieldElement]:
+    if len(vals) == 1:
+        return vals
+    L = _fft_field(vals[::2], roots_of_unity[::2])
+    R = _fft_field(vals[1::2], roots_of_unity[::2])
+    o = [BLSFieldElement(0) for _ in vals]
+    for i, (x, y) in enumerate(zip(L, R)):
+        y_times_root = (int(y) * int(roots_of_unity[i])) % BLS_MODULUS
+        o[i] = BLSFieldElement((int(x) + y_times_root) % BLS_MODULUS)
+        o[i + len(L)] = BLSFieldElement((int(x) - y_times_root + BLS_MODULUS) % BLS_MODULUS)
+    return o
```

Comment:
The current curve operation is basically pure Python. To use arkworks binding, the input `roots_of_unity` should be 
in encoded bytes format. Also, many new root of unity preset values should be changed to encoded bytes format to avoid having bigint in yaml files.